### PR TITLE
Fix flaky TestPositionsAppender.testConsecutiveBuilds

### LINKED ISSUE #12875: Flaky TestPositionsAppender.testConsecutiveBuilds
We can reproduce by increasing `invocationCount`, 10 is enough in my environment. 

```sh
com.google.common.base.VerifyException: position count 1 too small to have at least one null with rate 0.5

	at com.google.common.base.Verify.verify(Verify.java:417)
	at io.trino.block.BlockAssertions.chooseNullPositions(BlockAssertions.java:362)
	at io.trino.block.BlockAssertions.generateListWithNulls(BlockAssertions.java:351)
	at io.trino.block.BlockAssertions.createRandomLongsBlock(BlockAssertions.java:301)
	at io.trino.block.BlockAssertions.createRandomBlockForType(BlockAssertions.java:151)
	at io.trino.block.BlockAssertions.createRandomBlockForNestedType(BlockAssertions.java:223)
	at io.trino.block.BlockAssertions.createRandomBlockForType(BlockAssertions.java:195)
	at io.trino.operator.output.TestPositionsAppender.testConsecutiveBuilds(TestPositionsAppender.java:188)
	at jdk.internal.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)
```
https://github.com/trinodb/trino/runs/6903695040

cc: @sopel39 @lukasz-stec

### PR DESCRIPTION
**Pull request by lukasz-stec**
## Description
Since for nested types createRandomBlockForNestedType
chooses the lengths of the nested values randomly,
there is a chance that the chosen value length is
not big enough for the given nullRate.
To fix this, we cap the nested values nullRate, possibly to 0,
so it matches the nested position count.

> Is this change a fix, improvement, new feature, refactoring, or other?

fix flaky test
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine tests
> How would you describe this change to a non-technical end user or system administrator?

fix flaky test
## Related issues, pull requests, and links

* Fixes #12875

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

Fixes https://github.com/trinodb/trino/pull/12888

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 1]
```diff
@@ -941,4 +941,13 @@ private static void verifyNullRate(float nullRate)
     {
         verify(nullRate >= 0 && nullRate <= 1, "nullRate %s is not valid", nullRate);
     }
+
+    private static float capNullRate(float nullRate, int positionCount)
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:945]: Why is the variable named `positionCount` ? For map and array isn't `offsets[positionCount]` the total length of data in the Block ?

Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:945]: > For map and array isn't offsets[positionCount] the total length of data in the Block ?

well yes, if by total length you mean position count of the data/value block

Comment (raunaqmorarka) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:945]: Could you call it `entryCount` or `valuesLength` ? At least that's the terminology used in AbstractArrayBlock or AbstractMapBlock to distinguish it from `positionCount`.
Also consider passing down `offsets` and `positionCount` separately to this method if it's always meant to be used in context of Map and Row

Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:945]: This method is not specific to `AbstractMapBlock` or `AbstractArrayBlock`. It returns the nullRate that will match exactly the number of null positions for the given position count.

**Comment by sopel39** [comment]
Why can't we just fix `chooseNullPositions` to accept 0 nulls?

**Comment by lukasz-stec** [comment]
> Why can't we just fix `chooseNullPositions` to accept 0 nulls?

It accepts 0 nulls already. it fails if the null rate is not 0 but the result would be empty. This is to guard against unintentional setup where non-zero `nullRate` generates 0 nulls.

**Comment by sopel39** [comment]
> It accepts 0 nulls already. it fails if the null rate is not 0 but the result would be empty. This is to guard against unintentional setup where non-zero nullRate generates 0 nulls.

With `capNullRate` it seems we are purposefully mitigating that check, hence why to keep it in the first place?

**Comment by lukasz-stec** [comment]
> > It accepts 0 nulls already. it fails if the null rate is not 0 but the result would be empty. This is to guard against unintentional setup where non-zero nullRate generates 0 nulls.
> 
> With `capNullRate` it seems we are purposefully mitigating that check, hence why to keep it in the first place?

Yes, but only in the nested type case for the value block.
I'm also fine with dropping this check.

**Comment by sopel39** [comment]
> do we need two? how would use this new method given there is a recursive call in createRandomBlockForNestedType?

I think it's cleaner than adding a method with check and then creating another method that mitigates that check. You already have two methods

**Comment by lukasz-stec** [comment]
> > do we need two? how would use this new method given there is a recursive call in createRandomBlockForNestedType?
> 
> I think it's cleaner than adding a method with check and then creating another method that mitigates that check. You already have two methods

there is a recursive call in createRandomBlockForNestedType

**Comment by lukasz-stec** [comment]
> > there is a recursive call in createRandomBlockForNestedType
> 
> Would it be possible to simply exclude test cases that generate 0 nulls?

after offline discussion I changed this fix to make the `BlockAssertions` deterministic

**Review by sopel39 — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 2]
```diff
@@ -129,15 +127,15 @@ public static DictionaryBlock createRandomDictionaryBlock(Block dictionary, int
         checkArgument(dictionary.getPositionCount() > 0, "dictionary position count %s is less than or equal to 0", dictionary.getPositionCount());
 
         int[] ids = IntStream.range(0, positionCount)
-                .map(i -> RANDOM.nextInt(dictionary.getPositionCount()))
+                .map(i -> random().nextInt(dictionary.getPositionCount()))
```
Comment (ksobolew) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:130]: Um, this actually uses a fresh `Random` instance per iteration. Each instance has the same seed, so each iteration will produce the same result. So this is kind of more deterministic than you wanted, I think :)

Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:130]: :facepalm: https://github.com/trinodb/trino/pull/13029 cc @sopel39

**Pull request merged**